LocalApi.py

```diff
@@ -0,0 +1,250 @@
+
+        Store the current chunk region and titles in a buffer, when we find a new chunk we know the beginning 
+        and end of the previous chunk so we can add them to the dictionary.
+        """
+        buffer_region_start = 0
+        buffer_titles = []
+        with open(self.index_file, "r") as index_file:
+            for line in index_file:
+                [line_region_start, _id, *split_name] = line.split(":")
+                line_article_name = ":".join(split_name).strip()
+                if line_region_start != buffer_region_start:
+                    for name in buffer_titles:
+                        self.articles[name] = (int(buffer_region_start), int(line_region_start))
+                    buffer_region_start = line_region_start
+                    buffer_titles = []
+                buffer_titles.append(line_article_name)
+        for name in buffer_titles:
+            self.articles[name] = (int(buffer_region_start), int(self.xml_file_size))
+
+    def load(self):
+        """
+        Load the wikipedia index file into memory.
+        If an articles.pickle.tmp exists, we simply load that (~12 seconds)
+        If the pickle file doesn't exist, we manually load the index by parsing the text file (~43 seconds)
+        """
+        try:
+            with open("articles.pickle.tmp", "rb") as articles_pickle_file:
+                log.debug("Using pickled article file")
+                self.articles = pickle.load(articles_pickle_file)
+        except IOError:
+            log.debug("Unable to use pickled articles, doing a manual load of the index file")
+            self.manual_load()
+            with open("articles.pickle.tmp", "wb") as articles_pickle_file:
+                log.debug("Saving articles as a pickled dump to speed up future loads")
+                pickle.dump(self.articles, articles_pickle_file)
+        self.article_titles = list(self.articles.keys())
+
+    def get_range(self, title):
+        """
+        Find the range of bytes for a given article based on the title.
+
+        Returns (start, end) in bytes of the bzipped xml file
+        """
+        return self.articles[title]
+
+    def page_exists(self, title):
+        """
+        Return True IFF the article exists in the dictionary
+        """
+        return title in self.article_titles
+
+    def get_random_title(self):
+        """
+        Return a random title from the list of articles
+        """
+        return random.choice(self.article_titles)
+
+class LocalWikipediaApi(IWikiApi):
+    def __init__(self, index_file, bz_xml_file):
+        self.index_file = WikipediaIndexFile(index_file, getsize(bz_xml_file))
+        self.bz_xml_file = bz_xml_file
+
+    def load(self):
+        """
+        Loads the index file into memory.
+        Will potentially block the main thread for a little under a minute.
+        Must be called before using this API.
```

Comment:
Why not call `load` from the constructor if you need to call it before using it?

Reply:
Because it takes ages to load. I would imagine we don't want object constructions to just randomly take a long time... It's much better if we have a separate method that we can call, with the knowledge that it might block